src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,714 @@
+        require(newCrossPrice >= PRICE_SCALE - MAX_CROSS_PRICE_DEVIATION, "ARM: cross price too low");
+        require(newCrossPrice <= PRICE_SCALE, "ARM: cross price too high");
+
+        // If the new cross price is lower than the current cross price
+        if (newCrossPrice < crossPrice) {
+            // Check there is not a significant amount of base assets in the ARM
+            require(IERC20(baseAsset).balanceOf(address(this)) < MIN_TOTAL_SUPPLY, "ARM: too many base assets");
+        }
+
+        crossPrice = newCrossPrice;
+
+        emit CrossPriceUpdated(newCrossPrice);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Liquidity Provider Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Preview the amount of shares that would be minted for a given amount of assets
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that would be minted
+    function previewDeposit(uint256 assets) external view returns (uint256 shares) {
+        shares = convertToShares(assets);
+    }
+
+    /// @notice deposit liquidity assets in exchange for liquidity provider (LP) shares.
+    /// The caller needs to have approved the contract to transfer the assets.
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that were minted
+    function deposit(uint256 assets) external returns (uint256 shares) {
+        shares = _deposit(assets, msg.sender);
+    }
+
+    /// @notice deposit liquidity assets in exchange for liquidity provider (LP) shares.
+    /// This function is restricted to the Zap contract.
+    /// @param assets The amount of liquidity assets to deposit
+    /// @param liquidityProvider The address of the liquidity provider
+    /// @return shares The amount of shares that were minted
+    function deposit(uint256 assets, address liquidityProvider) external returns (uint256 shares) {
+        require(msg.sender == zap, "Only Zap");
+
+        shares = _deposit(assets, liquidityProvider);
+    }
+
+    /// @dev Internal logic for depositing liquidity assets in exchange for liquidity provider (LP) shares.
+    function _deposit(uint256 assets, address liquidityProvider) internal returns (uint256 shares) {
+        // Calculate the amount of shares to mint after the performance fees have been accrued
+        // which reduces the available assets, and before new assets are deposited.
+        shares = convertToShares(assets);
+
+        // Transfer the liquidity asset from the sender to this contract
+        IERC20(liquidityAsset).transferFrom(msg.sender, address(this), assets);
+
+        // mint shares
+        _mint(liquidityProvider, shares);
+
+        // Add the deposited assets to the last available assets
+        lastAvailableAssets += SafeCast.toInt128(SafeCast.toInt256(assets));
```

Comment:
⚪️ In general we want to do computation and state changes before doing external write calls. We can move `lastAvailableAssets +=` up to before the transferFrom.

Reply:
Fix proposed in this commit [43efe49](https://github.com/OriginProtocol/arm-oeth/pull/33/commits/43efe4995ce6fddac9bd31f61a3fd80eb382d971).